File: fs/before_checksum/779__max-chunks-to-make-sorted-ii/cpp/max-chunks-to-make-sorted-ii_benchmark_v2.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

// Time:  O(n)
// Space: O(n)

// mono stack solution
class Solution {
public:
    int maxChunksToSorted(vector<int>& arr) {
        vector<int> increasing_stk;
        for (const int& num : arr) {
            int max_num = increasing_stk.empty() ? num : max(increasing_stk.back(), num);
            while (!increasing_stk.empty() && increasing_stk.back() > num) {
                increasing_stk.pop_back();
            }
            increasing_stk.push_back(max_num);
        }
        return static_cast<int>(increasing_stk.size());
    }
};

int main() {
    vector<vector<int>> tests = {
        {1, 2, 3, 4, 5},
        {5, 4, 3, 2, 1},
        {1, 1, 1, 1},
        {2, 1, 3, 4, 4},
        {2, 2, 1, 1, 3, 3},
        {-2, -1, -3, 4, 3},
        {0},
        {1, 0, 1, 3, 2},
        {1000000000, -1000000000, 0, 999999999, -999999999},
        {1, 3, 2, 2, 2}
    };

    volatile int sink = 0;
    const int iterations = 1;
    for (int iter = 0; iter < iterations; ++iter) {
        int checksum = 0;
        Solution sol;
        for (auto& t : tests) {
            checksum += sol.maxChunksToSorted(t);
        }
        sink += checksum;
        

    } std::cout << sink << std::endl;

    return 0;
}

File: fs/py_before_checksum/779__max-chunks-to-make-sorted-ii/max-chunks-to-make-sorted-ii_benchmark_v2.py
# Time:  O(n)
# Space: O(n)

# mono stack solution
class Solution(object):
    def maxChunksToSorted(self, arr):
        """
        :type arr: List[int]
        :rtype: int
        """
        result, increasing_stk = 0, []
        for num in arr:
            max_num = num if not increasing_stk else max(increasing_stk[-1], num)
            while increasing_stk and increasing_stk[-1] > num:
                increasing_stk.pop()
            increasing_stk.append(max_num)
        return len(increasing_stk)


# Time:  O(nlogn)
# Space: O(n)
class Solution2(object):
    def maxChunksToSorted(self, arr):
        """
        :type arr: List[int]
        :rtype: int
        """
        def compare(i1, i2):
            return arr[i1]-arr[i2] if arr[i1] != arr[i2] else i1-i2

        idxs = [i for i in xrange(len(arr))]
        result, max_i = 0, 0
        for i, v in enumerate(sorted(idxs, cmp=compare)):
            max_i = max(max_i, v)
            if max_i == i:
                result += 1
        return result


if __name__ == "__main__":
    tests = [
        [1, 2, 3, 4, 5],
        [5, 4, 3, 2, 1],
        [1, 1, 1, 1],
        [2, 1, 3, 4, 4],
        [2, 2, 1, 1, 3, 3],
        [-2, -1, -3, 4, 3],
        [0],
        [1, 0, 1, 3, 2],
        [1000000000, -1000000000, 0, 999999999, -999999999],
        [1, 3, 2, 2, 2]
    ]

    sink = 0
    iterations = 1
    for iter in xrange(iterations):
        checksum = 0
        sol = Solution()
        for t in tests:
            checksum += sol.maxChunksToSorted(t)
        sink += checksum
    print(sink)
    # print sink